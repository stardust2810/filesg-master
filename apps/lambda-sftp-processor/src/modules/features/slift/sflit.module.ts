import { Module } from '@nestjs/common';

import { AwsModule } from '../aws/aws.module';
import { SliftService } from './slift.service';

@Module({
  providers: [SliftService],
  imports: [AwsModule],
  exports: [SliftService],
})
export class SliftModule {}
