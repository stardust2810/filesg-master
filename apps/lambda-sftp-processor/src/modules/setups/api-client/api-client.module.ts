import { Module } from '@nestjs/common';

import { CoreServiceClientProvider, TransferServiceClientProvider } from './api-client.provider';

@Module({
  providers: [CoreServiceClientProvider, TransferServiceClientProvider],
  exports: [CoreServiceClientProvider, TransferServiceClientProvider],
})
export class ApiClientModule {}
