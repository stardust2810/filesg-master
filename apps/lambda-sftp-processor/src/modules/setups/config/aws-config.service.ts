import { Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { Expose } from 'class-transformer';
import { IsNumber, IsString } from 'class-validator';

@Injectable()
export class AWSConfigService {
  constructor(private configService: ConfigService<AWSEnvironmentVariables>) {}

  get region() {
    return this.configService.get('AWS_REGION', { infer: true })!;
  }

  get sqsSftpProcessor() {
    return this.configService.get('AWS_SQS_SFTP_PROCESSOR', { infer: true })!;
  }

  get s3SftpBucket() {
    return this.configService.get('AWS_S3_BUCKET_SFTP', { infer: true })!;
  }

  get sftpRoleArn() {
    return this.configService.get('AWS_IAM_SFTP_ROLE_ARN', { infer: true })!;
  }

  get assumeRoleSessionDuration() {
    return this.configService.get('AWS_STS_ASSUME_ROLE_SESSION_DURATION', { infer: true })!;
  }
}

export class AWSEnvironmentVariables {
  @Expose()
  @IsString()
  AWS_REGION: string;

  @Expose()
  @IsString()
  AWS_SQS_SFTP_PROCESSOR: string;

  @Expose()
  @IsString()
  AWS_S3_BUCKET_SFTP: string;

  @Expose()
  @IsString()
  AWS_IAM_SFTP_ROLE_ARN: string;

  @Expose()
  @IsNumber()
  AWS_STS_ASSUME_ROLE_SESSION_DURATION: number;
}
