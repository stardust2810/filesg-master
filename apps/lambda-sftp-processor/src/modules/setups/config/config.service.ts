import { Injectable } from '@nestjs/common';

import { AWSConfigService } from './aws-config.service';
import { SliftConfigService } from './slift-config.service';
import { SystemConfigService } from './system-config.service';

@Injectable()
export class FileSGConfigService {
  constructor(
    private systemConfigService: SystemConfigService,
    private awsConfigService: AWSConfigService,
    private sliftConfigService: SliftConfigService,
  ) {}

  get systemConfig() {
    return this.systemConfigService;
  }
  get awsConfig() {
    return this.awsConfigService;
  }

  get sliftConfig() {
    return this.sliftConfigService;
  }
}
