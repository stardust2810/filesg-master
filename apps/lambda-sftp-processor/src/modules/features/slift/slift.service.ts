import { createDir, execShellCmd, isFileExists, LogMethod, moveFile, writeToFile } from '@filesg/backend-common';
import { Injectable, Logger } from '@nestjs/common';

import {
  SliftCertGenerationException,
  SliftDecryptException,
  SliftFileTypeException,
  SliftMissingException,
} from '../../../common/custom-exceptions';
import { SLIFT_ENCRYPTED_FILE_EXT, SLIFT_RECEIVER_PFX_BASE64, SLIFT_RECEIVER_PFX_PASSWORD } from '../../../const';
import { FileSGConfigService } from '../../setups/config/config.service';
import { SmService } from '../aws/sm.service';

@Injectable()
export class SliftService {
  private logger = new Logger(SliftService.name);
  private certDir: string;
  private privateCertPath: string;
  private privateCertBase64: string;
  private privateCertPassword: string;

  constructor(private readonly smService: SmService, private readonly fileSGConfigService: FileSGConfigService) {}

  @LogMethod()
  public async init(workingDir: string) {
    this.certDir = `${workingDir}/certs`;
    this.privateCertPath = `${this.certDir}/receiver.pfx`;
    await createDir(this.certDir);
    await this.generatePrivateKeyAndSaveToDisk();
    this.verifySliftScriptPresent();
  }

  @LogMethod()
  public async decrypt(srcPath: string, destPath: string): Promise<void> {
    if (!this.privateCertPassword) {
      this.logger.log(`Getting certifcate private key password from secrets manager.`);
      this.privateCertPassword = await this.smService.getSecretValue(SLIFT_RECEIVER_PFX_PASSWORD);
    }

    const { sliftDir } = this.fileSGConfigService.sliftConfig;

    this.logger.log(`Checking if the file extension is of type ${SLIFT_ENCRYPTED_FILE_EXT} before decryption`);
    if (srcPath.slice(-SLIFT_ENCRYPTED_FILE_EXT.length) !== SLIFT_ENCRYPTED_FILE_EXT) {
      throw new SliftFileTypeException();
    }

    const decryptedFilePath = srcPath.slice(0, -SLIFT_ENCRYPTED_FILE_EXT.length);

    try {
      this.logger.log(`Decrypting file ${decryptedFilePath}`);
      await execShellCmd(`cd ${sliftDir} && ./run.sh -d -pfx ${this.privateCertPath} ${this.privateCertPassword} ${srcPath}`);
    } catch (err) {
      throw new SliftDecryptException(srcPath, err as Error);
    }

    this.logger.log(`Copying decrypted file from ${decryptedFilePath} to ${destPath}`);
    await moveFile(decryptedFilePath, destPath);
  }

  // =============================================================================
  // Private functions
  // =============================================================================
  private async generatePrivateKeyAndSaveToDisk() {
    if (!this.privateCertBase64) {
      this.logger.log(`Getting certifcate private key from secrets manager.`);
      this.privateCertBase64 = await this.smService.getSecretValue(SLIFT_RECEIVER_PFX_BASE64);
    }

    try {
      this.logger.log(`Trying to generate the receiver pfx file from base64`);

      const binary = Buffer.from(this.privateCertBase64, 'base64');
      await writeToFile(this.privateCertPath, binary);

      this.logger.log(`Successfully created cretificate file with extension .pfx at location ${this.privateCertPath}`);
    } catch (err) {
      throw new SliftCertGenerationException(err as Error);
    }
  }

  private verifySliftScriptPresent() {
    const { sliftDir } = this.fileSGConfigService.sliftConfig;
    const sliftScriptLocation = `${sliftDir}/run.sh`;
    if (!isFileExists(sliftScriptLocation)) {
      throw new SliftMissingException(sliftDir);
    }
    this.logger.log(`Found slift script at location ${sliftScriptLocation}`);
  }
}
