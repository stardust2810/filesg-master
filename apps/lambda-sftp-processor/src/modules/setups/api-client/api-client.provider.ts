import { Provider } from '@nestjs/common';
import axios from 'axios';

import { CORE_API_CLIENT_PROVIDER, TRANSFER_API_CLIENT_PROVIDER } from '../../../const/index';
import { FileSGConfigService } from '../config/config.service';

export const CoreServiceClientProvider: Provider = {
  provide: CORE_API_CLIENT_PROVIDER,
  inject: [FileSGConfigService],
  useFactory: async (fileSGConfigService: FileSGConfigService) => {
    const { coreServiceUrl } = fileSGConfigService.systemConfig;

    return axios.create({
      baseURL: coreServiceUrl,
    });
  },
};

export const TransferServiceClientProvider: Provider = {
  provide: TRANSFER_API_CLIENT_PROVIDER,
  inject: [FileSGConfigService],
  useFactory: async (fileSGConfigService: FileSGConfigService) => {
    const { transferServiceUrl } = fileSGConfigService.systemConfig;

    return axios.create({
      baseURL: transferServiceUrl,
    });
  },
};
