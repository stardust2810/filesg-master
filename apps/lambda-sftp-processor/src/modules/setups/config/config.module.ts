import { Global, Module } from '@nestjs/common';
import { ConfigModule } from '@nestjs/config';

import { AWSConfigService } from './aws-config.service';
import { FileSGConfigService } from './config.service';
import { SliftConfigService } from './slift-config.service';
import { SystemConfigService } from './system-config.service';

@Global()
@Module({
  imports: [ConfigModule],
  providers: [SystemConfigService, AWSConfigService, FileSGConfigService, SliftConfigService],
  exports: [FileSGConfigService],
})
export class FileSGConfigModule {}
