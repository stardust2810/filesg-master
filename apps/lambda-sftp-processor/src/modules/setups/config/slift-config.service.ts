import { Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { Expose } from 'class-transformer';
import { IsNotEmpty, IsString } from 'class-validator';

@Injectable()
export class SliftConfigService {
  constructor(private configService: ConfigService<SliftEnvironmentVariables>) {}

  get sliftDir() {
    return this.configService.get('SLIFT_DIR', { infer: true })!;
  }
}

export class SliftEnvironmentVariables {
  @Expose()
  @IsString()
  @IsNotEmpty()
  SLIFT_DIR: string;
}
