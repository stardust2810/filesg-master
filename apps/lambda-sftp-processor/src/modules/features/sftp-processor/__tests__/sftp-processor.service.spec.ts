import { FsHelper } from '@filesg/backend-common';
import { COMPONENT_ERROR_CODE } from '@filesg/common';
import { UnzipService } from '@filesg/zipper';
import { HttpStatus } from '@nestjs/common';
import { Test, TestingModule } from '@nestjs/testing';
import FileType from 'file-type';
import fs from 'fs';
import { cloneDeep } from 'lodash';

import {
  CreateTransactionException,
  ExtraFileException,
  InvalidFilePathFormatAgencyPasswordException,
  ProcessMessageErrorException,
  RequestTimeoutException,
  UnsupportedFileTypeException,
  UploadTransactionFilesException,
} from '../../../../common/custom-exceptions';
import { ExtraFileAgencyPasswordException } from '../../../../common/custom-exceptions';
import { DuplicateEntryAgencyPasswordException } from '../../../../common/custom-exceptions';
import { CORE_API_CLIENT_PROVIDER, SIDECAR_FILES_INFO, TRANSFER_API_CLIENT_PROVIDER } from '../../../../const';
import { mockCoreServiceApiClient, mockTransferServiceApiClient } from '../../../setups/api-client/__mocks__/api-client.mock';
import { mockFileSGConfigService } from '../../../setups/config/__mocks__/config.service.mock';
import { FileSGConfigService } from '../../../setups/config/config.service';
import { mockS3Service } from '../../aws/__mocks__/aws-s3.service.mock';
import { mockMessageId, mockMsgReceiptHandle, mockSqsService } from '../../aws/__mocks__/aws-sqs.service.mock';
import { S3Service } from '../../aws/s3.service';
import { SqsService } from '../../aws/sqs.service';
import { mockSliftService } from '../../slift/__mocks__/slift.service.mock';
import { SliftService } from '../../slift/slift.service';
import {
  mockCreateTransactionPayload,
  mockCreateTransactionResponse,
  mockFilePath,
  mockFileTypeResult,
  mockJwt,
  mockLambdaSqsMessage,
  mockPaths,
  mockS3Key,
  TestSftpProcessorService,
} from '../__mocks__/sftp-processor.service.mock';
import { mockSidecarData, mockSidecarFileService, mockUnzipService } from '../__mocks__/sidecar-file.service.mock';
import { SidecarFileService } from '../sidecar-file.service';

describe('SftpProcessorService', () => {
  let service: TestSftpProcessorService;

  beforeEach(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [
        TestSftpProcessorService,
        { provide: S3Service, useValue: mockS3Service },
        { provide: SqsService, useValue: mockSqsService },
        { provide: SidecarFileService, useValue: mockSidecarFileService },
        { provide: UnzipService, useValue: mockUnzipService },
        { provide: SliftService, useValue: mockSliftService },
        { provide: FileSGConfigService, useValue: mockFileSGConfigService },
        { provide: CORE_API_CLIENT_PROVIDER, useValue: mockCoreServiceApiClient },
        { provide: TRANSFER_API_CLIENT_PROVIDER, useValue: mockTransferServiceApiClient },
      ],
    }).compile();

    service = module.get<TestSftpProcessorService>(TestSftpProcessorService);
  });

  afterEach(() => {
    jest.resetAllMocks();
    jest.restoreAllMocks();
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  describe('run', () => {
    it('ProcessMessageErrorException should be thrown when body is empty', async () => {
      await expect(service.run({ ...mockLambdaSqsMessage, body: '' })).rejects.toThrowError(ProcessMessageErrorException);
    });

    it('ProcessMessageErrorException should be thrown when body cannot be parsed', async () => {
      await expect(service.run({ ...mockLambdaSqsMessage, body: 'non json body' })).rejects.toThrowError(ProcessMessageErrorException);
    });

    it('ProcessMessageErrorException should be thrown when there are no record', async () => {
      await expect(service.run({ ...mockLambdaSqsMessage, body: '{"Records":[]}' })).rejects.toThrowError(ProcessMessageErrorException);
    });

    it('errorHandler should not be called where there is no processing error', async () => {
      jest.spyOn(service, 'generateFilePaths').mockReturnValue(mockPaths('mock-uuid'));
      jest.spyOn(service, 'preProcess').mockImplementation();
      jest.spyOn(service, 'parseAndValidateCsv').mockResolvedValue(mockSidecarData);
      jest.spyOn(service, 'process').mockImplementation();
      jest.spyOn(service, 'postProcess').mockImplementation();
      jest.spyOn(service, 'errorHandler').mockImplementation();
      jest.spyOn(service, 'clean').mockImplementation();

      await expect(service.run(mockLambdaSqsMessage)).resolves.not.toThrow();
      expect(service.errorHandler).toBeCalledTimes(0);
      expect(service.clean).toBeCalledTimes(1);
    });

    it('errorHandler and clean should be called where there is processing error', async () => {
      jest.spyOn(service, 'generateFilePaths').mockReturnValue(mockPaths('mock-uuid'));
      jest.spyOn(service, 'preProcess').mockImplementation();
      jest.spyOn(service, 'parseAndValidateCsv').mockResolvedValue(mockSidecarData);
      jest.spyOn(service, 'process').mockImplementation(() => {
        throw new Error();
      });
      jest.spyOn(service, 'postProcess').mockImplementation();
      jest.spyOn(service, 'errorHandler').mockImplementation();
      jest.spyOn(service, 'clean').mockImplementation();

      await expect(service.run(mockLambdaSqsMessage)).resolves.not.toThrow();
      expect(service.errorHandler).toBeCalledTimes(1);
      expect(service.clean).toBeCalledTimes(1);
    });
  });

  describe('preProcess', () => {
    it('the underlying methods are called with correct args', async () => {
      const paths = mockPaths('mock-uuid');

      jest.spyOn(FsHelper, 'createDir').mockImplementation();
      jest.spyOn(FsHelper, 'rmFile').mockImplementation();
      jest.spyOn(fs, 'createReadStream').mockImplementation();

      await service.preProcess(paths);
      expect(FsHelper.createDir).toBeCalledWith(paths.localWorkingDir);
      expect(mockSliftService.init).toBeCalledWith(paths.localWorkingDir);
      expect(mockS3Service.downloadFileFromSftpBucketToDisk).toBeCalledWith(paths.s3Key, paths.encryptedZip);
      expect(mockSliftService.decrypt).toBeCalledWith(paths.encryptedZip, paths.plainZip);
      expect(FsHelper.rmFile).toBeCalledWith(paths.encryptedZip);
      expect(mockUnzipService.unzipToDisk).toBeCalledTimes(1);
    });
  });

  describe('parseAndValidateCsv', () => {
    it('the underlying methods are called with correct args', async () => {
      const paths = mockPaths('mock-uuid');

      mockSidecarFileService.parseSidecarFiles.mockResolvedValueOnce(mockSidecarData);
      jest.spyOn(service, 'checkForMissingIssuanceFiles').mockImplementation();
      jest.spyOn(service, 'checkForExtraIssuanceFiles').mockImplementation();
      jest.spyOn(service, 'detectAndPopulateMimeType').mockImplementation();
      jest.spyOn(service, 'validateFilePathInsideAgencyPasswordSidecar').mockImplementation();

      await expect(service.parseAndValidateCsv(paths)).resolves.toEqual(mockSidecarData);
      expect(mockSidecarFileService.checkSidecarFilesExistsOrThrow).toBeCalledWith(paths.extractedZipDir);
      expect(mockSidecarFileService.parseSidecarFiles).toBeCalledWith(paths.extractedZipDir);
      expect(service.checkForMissingIssuanceFiles).toBeCalledWith(paths, mockSidecarData);
      expect(service.checkForExtraIssuanceFiles).toBeCalledWith(paths, mockSidecarData);
      expect(service.detectAndPopulateMimeType).toBeCalledWith(paths, mockSidecarData);
      expect(service.validateFilePathInsideAgencyPasswordSidecar).toBeCalledWith(paths, mockSidecarData);
    });
  });

  describe('process', () => {
    it('the underlying methods are called with correct args', async () => {
      const paths = mockPaths('mock-uuid');

      jest.spyOn(service, 'createTransaction').mockResolvedValueOnce(mockCreateTransactionResponse);
      jest.spyOn(service, 'uploadTransactionFiles').mockImplementation();
      jest.spyOn(fs, 'createReadStream').mockImplementation();

      await service.process(paths, mockSidecarData);

      // unable to test with toBeCalledWith due to ReadStream
      expect(mockS3Service.uploadFilesToSftpBucket).toBeCalledTimes(1);
      expect(service.createTransaction).toBeCalledWith(mockSidecarData);
      expect(service.uploadTransactionFiles).toBeCalledWith(mockJwt, mockSidecarData, paths.s3WorkingDir);
    });
  });

  describe('postProcess', () => {
    it('deleteFileFromSftpBucket and deleteMessageInQueueSftpProcessor should be called with correct args', async () => {
      const paths = mockPaths('mock-uuid');

      await service.postProcess(paths, mockLambdaSqsMessage);

      expect(mockS3Service.deleteFileFromSftpBucket).toBeCalledWith(paths.s3Key);
      expect(mockSqsService.deleteMessageInQueueSftpProcessor).toBeCalledWith(
        mockLambdaSqsMessage.messageId,
        mockLambdaSqsMessage.receiptHandle,
      );
    });
  });

  describe('clean', () => {
    it('rmDir and deleteFilesByPrefixFromSftpBucket should be called with correct args', async () => {
      const paths = mockPaths('mock-uuid');

      jest.spyOn(FsHelper, 'rmDir').mockResolvedValueOnce(undefined);
      mockS3Service.deleteFilesByPrefixFromSftpBucket.mockResolvedValueOnce(undefined);

      await service.clean(paths);

      expect(FsHelper.rmDir).toBeCalledWith(paths.localWorkingDir);
      expect(mockS3Service.deleteFilesByPrefixFromSftpBucket).toBeCalledWith(paths.s3WorkingDir);
    });
  });

  describe('generateFilePaths', () => {
    it('file paths should be generated correctly', () => {
      const generatedPaths = service.generateFilePaths(mockS3Key);
      const generatedUuid = generatedPaths.processUuid;

      expect(generatedPaths).toEqual(mockPaths(generatedUuid));
    });
  });

  describe('deriveFilenames', () => {
    it('filename should be derived correctly', () => {
      const paths = mockPaths('mock-uuid');

      expect(service.deriveFilenames(mockS3Key)).toEqual({
        encryptedFileName: paths.encryptedFileName,
        plainFileName: paths.plainFileName,
      });
    });
  });

  describe('checkForMissingIssuanceFiles', () => {
    it('no exception should be thrown if there are no missing issuance files', async () => {
      const paths = mockPaths('mock-uuid');

      jest.spyOn(FsHelper, 'checkForMissingFilesInDir').mockResolvedValueOnce([]);

      await expect(service.checkForMissingIssuanceFiles(paths, mockSidecarData)).resolves.not.toThrow();
    });

    it('MissingFileException should be throw if there are any missing file', async () => {
      const paths = mockPaths('mock-uuid');

      jest.spyOn(FsHelper, 'checkForMissingFilesInDir').mockResolvedValueOnce(['mock-missing-file.pdf']);

      await expect(service.checkForMissingIssuanceFiles(paths, mockSidecarData)).rejects.toThrow();
    });
  });

  describe('checkForExtraIssuanceFiles', () => {
    it('no exception should be thrown if there are no extra files', async () => {
      const paths = mockPaths('mock-uuid');

      jest
        .spyOn(FsHelper, 'listDirContent')
        .mockResolvedValueOnce([...mockSidecarData.files.map((file) => file.name), ...SIDECAR_FILES_INFO.map((info) => info.name)]);

      await expect(service.checkForExtraIssuanceFiles(paths, mockSidecarData)).resolves.not.toThrow();
    });

    it('ExtraFileException should be thrown if there are extra files', async () => {
      const paths = mockPaths('mock-uuid');

      jest
        .spyOn(FsHelper, 'listDirContent')
        .mockResolvedValueOnce([
          ...mockSidecarData.files.map((file) => file.name),
          ...SIDECAR_FILES_INFO.map((info) => info.name),
          'extra-mock-file.pdf',
        ]);

      await expect(service.checkForExtraIssuanceFiles(paths, mockSidecarData)).rejects.toThrowError(ExtraFileException);
    });
  });

  describe('detectAndPopulateMimeType', () => {
    it('mimeType should be populated correctly', async () => {
      const paths = mockPaths('mock-uuid');
      const clonedMockSidecarData = cloneDeep(mockSidecarData);

      jest.spyOn(FileType, 'fromFile').mockResolvedValue(mockFileTypeResult);

      await expect(service.detectAndPopulateMimeType(paths, clonedMockSidecarData)).resolves.not.toThrow();
      expect(clonedMockSidecarData.files[0].mimeType).toEqual('application/pdf');
      expect(clonedMockSidecarData.files[1].mimeType).toEqual('application/pdf');
    });

    it('UnsupportedFileTypeException should be thrown if mimeType cannot be detected', async () => {
      const paths = mockPaths('mock-uuid');
      const clonedMockSidecarData = cloneDeep(mockSidecarData);

      jest.spyOn(FileType, 'fromFile').mockResolvedValue(undefined);

      await expect(service.detectAndPopulateMimeType(paths, clonedMockSidecarData)).rejects.toThrowError(UnsupportedFileTypeException);
    });

    it('UnsupportedFileTypeException should be thrown if mimeType is unsupported', async () => {
      const paths = mockPaths('mock-uuid');
      const clonedMockSidecarData = cloneDeep(mockSidecarData);

      jest.spyOn(FileType, 'fromFile').mockResolvedValue({ ext: 'mp4', mime: 'video/mp4' });

      await expect(service.detectAndPopulateMimeType(paths, clonedMockSidecarData)).rejects.toThrowError(UnsupportedFileTypeException);
    });
  });

  describe('validateFilePathInsideAgencyPasswordSidecar', () => {
    it('should throw error if the agency password sidecar file has record and not a corresponding physical file', async () => {
      const paths = mockPaths('mock-uuid');
      const clonedMockSidecarData = cloneDeep(mockSidecarData);
      clonedMockSidecarData.agencyPassword = [{ filePath: 'non-existing-file.zip/folder/filename.pdf', password: 'P@ssw0rd' }];
      jest.spyOn(fs, 'existsSync').mockReturnValue(false);

      expect(service.validateFilePathInsideAgencyPasswordSidecar(paths, clonedMockSidecarData)).rejects.toThrowError(
        ExtraFileAgencyPasswordException,
      );
    });

    it('should throw error if the agency password sidecar filepath has invalid paths', async () => {
      const paths = mockPaths('mock-uuid');
      const clonedMockSidecarData = cloneDeep(mockSidecarData);
      clonedMockSidecarData.agencyPassword = [
        { filePath: 'existing-file.zip/folder/', password: 'P@ssw0rd' },
        { filePath: 'existing-file.zip', password: 'P@ssw0rd' },
      ];
      jest.spyOn(fs, 'existsSync').mockReturnValue(true);

      expect(service.validateFilePathInsideAgencyPasswordSidecar(paths, clonedMockSidecarData)).rejects.toThrowError(
        InvalidFilePathFormatAgencyPasswordException,
      );
    });

    it('should throw error if the agency password sidecar filepath has duplicate entry', async () => {
      const paths = mockPaths('mock-uuid');
      const clonedMockSidecarData = cloneDeep(mockSidecarData);
      clonedMockSidecarData.agencyPassword = [
        { filePath: 'existing-file.zip/folder/filename.pdf', password: 'P@ssw0rd' },
        { filePath: 'existing-file.zip/folder/filename.pdf', password: 'P@ssw0rd' },
      ];
      jest.spyOn(fs, 'existsSync').mockReturnValue(true);

      expect(service.validateFilePathInsideAgencyPasswordSidecar(paths, clonedMockSidecarData)).rejects.toThrowError(
        DuplicateEntryAgencyPasswordException,
      );
    });
  });

  describe('createTransaction', () => {
    it('coreServiceClient is called with correct args', async () => {
      mockCoreServiceApiClient.post.mockResolvedValueOnce({ data: { data: mockCreateTransactionResponse } });
      jest.spyOn(service, 'getCreateTransactionPayload').mockReturnValue(mockCreateTransactionPayload);

      await expect(service.createTransaction(mockSidecarData)).resolves.toEqual(mockCreateTransactionResponse);
      expect(mockCoreServiceApiClient.post).toBeCalledWith('v2/transaction/file/client', mockCreateTransactionPayload, {
        headers: {
          'x-client-id': mockSidecarData.transactions[0].clientId,
          'x-client-secret': mockSidecarData.transactions[0].clientSecret,
        },
      });
    });

    it('CreateTransactionException should be thrown when api throws error', async () => {
      mockCoreServiceApiClient.post.mockRejectedValueOnce({ originErrorMsg: 'originErrorMsg', response: { data: 'errorData' } });

      await expect(service.createTransaction(mockSidecarData)).rejects.toThrowError(CreateTransactionException);
    });

    it('should throw RequestTimeoutException when api return status of 408', async () => {
      mockCoreServiceApiClient.post.mockRejectedValueOnce({
        originErrorMsg: 'originErrorMsg',
        response: { data: 'errorData', status: HttpStatus.REQUEST_TIMEOUT },
      });

      await expect(service.createTransaction(mockSidecarData)).rejects.toThrowError(RequestTimeoutException);
    });
  });

  describe('uploadTransactionFiles', () => {
    it('transferServiceClient is called with correct args', async () => {
      const paths = mockPaths('mock-uuid');
      const mockUploadTransactionFilesPayload = { files: [] };

      mockTransferServiceApiClient.post.mockResolvedValueOnce('mock return');
      jest.spyOn(service, 'getUploadTransactionFilesPayload').mockReturnValue(mockUploadTransactionFilesPayload);

      await expect(service.uploadTransactionFiles(mockJwt, mockSidecarData, paths.s3WorkingDir)).resolves.toEqual('mock return');
      expect(mockTransferServiceApiClient.post).toBeCalledWith('v1/file-upload', mockUploadTransactionFilesPayload, {
        headers: {
          Authorization: `Bearer ${mockJwt}`,
        },
      });
    });

    it('UploadTransactionFilesException should be thrown when api throws error', async () => {
      mockTransferServiceApiClient.post.mockRejectedValueOnce({ originErrorMsg: 'originErrorMsg', response: { data: 'errorData' } });

      const paths = mockPaths('mock-uuid');

      await expect(service.uploadTransactionFiles('mockJwt', mockSidecarData, paths.s3WorkingDir)).rejects.toThrow(
        UploadTransactionFilesException,
      );
    });

    it('should throw RequestTimeoutException when api return status of 408', async () => {
      mockTransferServiceApiClient.post.mockRejectedValueOnce({
        originErrorMsg: 'originErrorMsg',
        response: { data: 'errorData', status: HttpStatus.REQUEST_TIMEOUT },
      });

      const paths = mockPaths('mock-uuid');

      await expect(service.uploadTransactionFiles('mockJwt', mockSidecarData, paths.s3WorkingDir)).rejects.toThrow(RequestTimeoutException);
    });
  });

  describe('getCreateTransactionPayload', () => {
    it('data should be transformed correctly', () => {
      // using snapshot as the output is long
      expect(service.getCreateTransactionPayload(mockSidecarData)).toMatchSnapshot();
    });
  });

  describe('getUploadTransactionFilesPayload', () => {
    it('data should be transformed correctly', () => {
      const paths = mockPaths('mock-uuid');
      const { s3SftpBucket } = mockFileSGConfigService.awsConfig;

      expect(service.getUploadTransactionFilesPayload(mockSidecarData, paths.s3WorkingDir)).toEqual({
        files: [
          {
            fileName: mockSidecarData.files[0].name,
            isOA: false,
            s3FileData: {
              bucketName: s3SftpBucket,
              key: `${paths.s3WorkingDir}/${mockSidecarData.files[0].name}`,
            },
          },
          {
            fileName: mockSidecarData.files[1].name,
            isOA: false,
            s3FileData: {
              bucketName: s3SftpBucket,
              key: `${paths.s3WorkingDir}/${mockSidecarData.files[1].name}`,
            },
          },
        ],
      });
    });
  });

  describe('errorHandler', () => {
    it('deleteMessageInQueueSftpProcessor shoud not be called if error is retryable', () => {
      service.errorHandler(
        new ProcessMessageErrorException(COMPONENT_ERROR_CODE.SFTP_PROCESSOR_SERVICE, 'dummy'),
        mockFilePath,
        mockMessageId,
        mockMsgReceiptHandle,
      );

      expect(mockSqsService.deleteMessageInQueueSftpProcessor).toBeCalledTimes(0);
    });

    it('deleteMessageInQueueSftpProcessor shoud be called if error is not retryable', () => {
      service.errorHandler(
        new ExtraFileException(COMPONENT_ERROR_CODE.SFTP_PROCESSOR_SERVICE, ['mock-extra-file.pdf']),
        mockFilePath,
        mockMessageId,
        mockMsgReceiptHandle,
      );

      expect(mockSqsService.deleteMessageInQueueSftpProcessor).toBeCalledTimes(1);
    });
  });
});
