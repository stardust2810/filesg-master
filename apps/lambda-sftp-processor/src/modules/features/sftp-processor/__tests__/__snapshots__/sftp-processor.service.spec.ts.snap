// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SftpProcessorService getCreateTransactionPayload data should be transformed correctly 1`] = `
Object {
  "application": Object {
    "externalRefId": "mockApplicationExternalRefId",
    "type": "mockApplicationType",
  },
  "files": Array [
    Object {
      "checksum": "d2e5099b3cda2727f5f4df356ee3061b695113a311f11117de8c11713aa901b4",
      "deleteAt": "2024-12-31",
      "expiry": "2024-12-01",
      "isPasswordEncryptionRequired": true,
      "metadata": Object {
        "hello": "world",
      },
      "name": "mockFileName1.pdf",
    },
    Object {
      "checksum": "d2e5099b3cda2727f5f4df356ee3061b695113a311f11117de8c11713aa901b4",
      "deleteAt": "2024-12-31",
      "expiry": "2024-12-01",
      "isPasswordEncryptionRequired": true,
      "metadata": undefined,
      "name": "mockFileName2.pdf",
    },
  ],
  "transaction": Object {
    "acknowledgementTemplateUuid": "mockAcknowledgementTemplateId",
    "creationMethod": "sftp",
    "customAgencyMessage": Object {
      "notifications": undefined,
      "transaction": Object {
        "templateId": "transactioncustommessagetemplate-1655625225500-9c72b9eac3dc42f6",
        "templateInput": undefined,
      },
    },
    "isAcknowledgementRequired": true,
    "name": "mockTransactionName",
    "recipients": Array [
      Object {
        "contact": "+6598765432",
        "dob": "1999-01-01",
        "email": "test1@dummy.com",
        "metadata": Object {
          "hello": "world",
        },
        "name": "Mock Full Name 1",
        "uin": "S1000000I",
      },
      Object {
        "contact": "+6598111111",
        "dob": "1999-00-00",
        "email": "test2@dummy.com",
        "metadata": undefined,
        "name": "Mock Full Name 2",
        "uin": "S1000001G",
      },
    ],
    "type": "upload_transfer",
  },
}
`;
